### PR TITLE
Allow compatible shapes in FW Node, not only equal

### PR DESCRIPTION
**Pull request by mvafin**
### Details:
 - *Allow compatible shapes in FW Node, not only equal*

### Tickets:
 - *None*

### CODE REVIEW

File: src/core/src/op/util/framework_node.cpp [hunk 1]
```diff
@@ -149,17 +149,15 @@ void ov::op::util::FrameworkNode::validate_and_infer_types() {
...
                 reset_output_shape_to_dynamic = true;
             } else {
-                NODE_VALIDATION_CHECK(this,
-                                      std::get<0>(m_inputs_desc[i]).compatible(input_pshape) &&
-                                          std::get<1>(m_inputs_desc[i]).compatible(input_type),
-                                      get_error_message());
+                NODE_VALIDATION_CHECK(
+                    this,
+                    orig_input_pshape.compatible(input_pshape) && std::get<1>(m_inputs_desc[i]).compatible(input_type),
+                    get_error_message());
```
Comment (ilya-lavrenov) [inline src/core/src/op/util/framework_node.cpp:160]: should we add a test to fix the new behavior?

Comment (mvafin) [reply src/core/src/op/util/framework_node.cpp:160]: Yep, added

**Review by eaidova — APPROVED** [review]

File: src/core/tests/type_prop/framework_node.cpp [hunk 2]
```diff
@@ -50,3 +50,48 @@ TEST(type_prop, framework_node) {
...
 }
+
+TEST(type_prop, dynamic_framework_node_with_dynamic_input) {
+    auto param = std::make_shared<ov::opset8::Parameter>(ov::element::dynamic, ov::PartialShape::dynamic());
+    auto f_node = std::make_shared<ov::op::util::FrameworkNode>(ov::OutputVector{param});
+
+    // Set partially dynamic shape
+    param->set_partial_shape(ov::PartialShape{ov::Dimension::dynamic(), 64});
+    param->validate_and_infer_types();
+
```
Comment (rkazants) [inline src/core/tests/type_prop/framework_node.cpp:61]: Also, you can add testing with determined input type like `f32`, etc.

Comment (mvafin) [reply src/core/tests/type_prop/framework_node.cpp:61]: Such test already exist.

Comment (rkazants) [reply src/core/tests/type_prop/framework_node.cpp:61]: this is about not static shapes, question is about the element type:)

**Review by rkazants — APPROVED** [review]

**Comment by slyalin** [comment]
@mvafin, give me a basic understanding of why we need these checks in FW node at all. Why cannot we accept any input shape regardless of old ones? FW node doesn't have any behavior encapsulated, so it always requires another code that defines how shapes are propagated and which input shapes are acceptable. I mean, does this or any other possible definition of `validate_and_infer_types` really make sense?

**Review by slyalin — APPROVED** [review]

**Pull request merged**